dashboard_server.py

```
@@ -0,0 +1,615 @@
+
+        # UMAP layout from distances (precomputed)
+        try:
+            reducer = umap.UMAP(n_components=2, metric='precomputed', n_neighbors=min(10, n-1), min_dist=0.2, random_state=random_state)
+            emb = reducer.fit_transform(D)
+            # Normalize to [0,1]
+            min_xy = emb.min(axis=0)
+            max_xy = emb.max(axis=0)
+            span = np.where((max_xy - min_xy) == 0, 1.0, (max_xy - min_xy))
+            norm = (emb - min_xy) / span
+        except Exception as e:
+            logging.getLogger(__name__).exception(f"UMAP failed, falling back to spring_layout")
```

Ruff (0.12.2)

f-string without any placeholders

Remove extraneous `f` prefix

(F541)